### PR TITLE
Feat/329 api integration tests v2

### LINKED ISSUE #329: Expand monitoring API test coverage (unit + integration)
## Context

PR #281 introduced monitoring integration tests that validate the `/metrics` (Prometheus) and `/api/v1/health` endpoints. A reviewer noted that the JSON REST API endpoints are only covered by unit tests in `http_server.rs` but lack integration-level validation against a running topology.

## Goal

Extend test coverage to exercise the full JSON REST API surface in both unit tests and integration tests, ensuring the responses contain correct, meaningful data from live SV2 topologies.

## Endpoints needing integration tests

All JSON API endpoints (`/`, `/api/v1/global`, `/api/v1/server`, `/api/v1/server/channels`, `/api/v1/clients`, `/api/v1/clients/{id}`, `/api/v1/clients/{id}/channels`, `/api/v1/sv1/clients`, `/api/v1/sv1/clients/{id}`) currently have unit tests only. `/api/v1/health` and `/metrics` are already covered by integration tests.

## Suggested approach

1. Add integration tests to `monitoring_integration.rs` that reuse existing topology setups and make HTTP requests to each JSON endpoint, asserting on status codes and key fields.
2. Add assertion helpers to `prometheus_metrics_assertions.rs` for JSON API responses.
3. Strengthen unit tests in `http_server.rs` to cover edge cases (pagination boundaries, invalid path params, partial data sources).

## Acceptance criteria

- Every JSON API endpoint has at least one integration test against a live topology
- Unit tests cover error/edge cases for pagination and missing data
- All existing tests continue to pass
- `cargo fmt`, `cargo clippy -D warnings` clean

### PR DESCRIPTION
**Pull request by gimballock**
Follow up to #373 which was merged right as new review comments came in, this PR cherry picks the final commit from the previous PR and addresses the final review comments.

I'll track the updates here according to the new comments in the old PR one by one and provide a summary here.

I have some updates here already but still need to update the review comments (will update this comment when that is done).

close #329

### CODE REVIEW

File: integration-tests/lib/prometheus_metrics_assertions.rs [hunk 1]
```diff
@@ -27,6 +42,267 @@ pub async fn fetch_api(monitoring_addr: SocketAddr, path: &str) -> String {
...
+/// Fetch a JSON API endpoint returning both the HTTP status code and the parsed body
+/// deserialized into the caller-specified type `T`.
+///
+/// Unlike `fetch_api_typed`, this does **not** panic on non-2xx responses, so it can be
+/// used to test error endpoints. For 404/error bodies, parametrize `T` with the
+/// production `ErrorResponse` type to keep the assertion fully typed.
+pub async fn fetch_api_with_status_typed<T: serde::de::DeserializeOwned + Send + 'static>(
+    monitoring_addr: SocketAddr,
+    path: &str,
+) -> (i32, T) {
```
Comment (bit-aloo) [inline integration-tests/lib/prometheus_metrics_assertions.rs]: Can you have these methods inside some MonitoringAPI abstraction. Keep them in separate for easy to review.

File: stratum-apps/src/monitoring/routes.rs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+// ── Top-level paths ─────────────────────────────────────────────────
+
+pub const ROOT: &str = "/";
+pub const METRICS: &str = "/metrics";
+
+// ── /api/v1 prefix and segments ─────────────────────────────────────
+
+/// Common prefix for all versioned JSON API endpoints.
+pub const API_V1_PREFIX: &str = "/api/v1";
```
Comment (bit-aloo) [inline stratum-apps/src/monitoring/routes.rs:17]: Shouldn't these also be used inside the monitoring crate?

Comment (gimballock) [reply stratum-apps/src/monitoring/routes.rs:17]: the path constants are now used consistently inside the monitoring crate, not just in downstream tests

File: integration-tests/lib/prometheus_metrics_assertions.rs (hunk 1 above)
Comment (gimballock) [reply integration-tests/lib/prometheus_metrics_assertions.rs]: Wrapped all network-touching helpers under a single MonitoringApi struct that holds the SocketAddr. 
Call sites now read as `api.fetch_typed(path)` and `api.poll_global_sv2_clients_gte(n, t)` instead of passing the address into a dozen free functions.

File: stratum-apps/src/monitoring/routes.rs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+// Segments under `API_V1_PREFIX` (used by the axum nested router).
+pub const SEG_HEALTH: &str = "/health";
+pub const SEG_GLOBAL: &str = "/global";
+pub const SEG_SERVER: &str = "/server";
+pub const SEG_SERVER_CHANNELS: &str = "/server/channels";
+pub const SEG_CLIENTS: &str = "/clients";
+pub const SEG_CLIENT_BY_ID: &str = "/clients/{client_id}";
+pub const SEG_CLIENT_CHANNELS: &str = "/clients/{client_id}/channels";
+pub const SEG_SV1_CLIENTS: &str = "/sv1/clients";
+pub const SEG_SV1_CLIENT_BY_ID: &str = "/sv1/clients/{client_id}";
```
Comment (bit-aloo) [inline stratum-apps/src/monitoring/routes.rs]: Why `SEG` prefix?

Comment (gimballock) [reply stratum-apps/src/monitoring/routes.rs]: just to mean its not a full path from the api root, it comes after API_V1_PREFIX (the /api/v1/ part), i made it more explicit with it's own namespace now

File: stratum-apps/src/monitoring/routes.rs (hunk 2 above)
Comment (bit-aloo) [reply stratum-apps/src/monitoring/routes.rs:17]: @GitGab19, can you take a look at this, considering it touches the monitoring crate?

File: stratum-apps/src/monitoring/http_server.rs [hunk 4]
```diff
@@ -290,93 +300,139 @@ impl MonitoringServer {
...
+        "Single Sv1 client (Translator Proxy only)".to_string(),
+    );
+    endpoints.insert(
+        routes::METRICS.to_string(),
+        "Prometheus metrics".to_string(),
+    );
+
+    Json(RootResponse {
+        service: "SRI Monitoring API".to_string(),
+        version: "0.1.0".to_string(),
```
Comment (bit-aloo) [inline stratum-apps/src/monitoring/http_server.rs]: This shouldn't be hardcoded?

File: stratum-apps/src/monitoring/routes.rs [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+pub const SV1_CLIENT_BY_ID_PATTERN: &str = "/api/v1/sv1/clients/{client_id}";
+
+/// Path for `/api/v1/clients/{id}` with the id substituted in.
+pub fn client_by_id(id: usize) -> String {
+    format!("/api/v1/clients/{}", id)
+}
+
+/// Path for `/api/v1/clients/{id}/channels` with the id substituted in.
+pub fn client_channels(id: usize) -> String {
+    format!("/api/v1/clients/{}/channels", id)
```
Comment (bit-aloo) [inline stratum-apps/src/monitoring/routes.rs]:
```suggestion
    format!("/api/v1/clients/{id}/channels")
```

File: stratum-apps/src/monitoring/routes.rs [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+/// Path for `/api/v1/sv1/clients/{id}` with the id substituted in.
+pub fn sv1_client_by_id(id: usize) -> String {
+    format!("/api/v1/sv1/clients/{}", id)
```
Comment (bit-aloo) [inline stratum-apps/src/monitoring/routes.rs]:
```suggestion
    format!("/api/v1/sv1/clients/{id}")
```

File: stratum-apps/src/monitoring/http_server.rs (hunk 4 above)
Comment (gimballock) [reply stratum-apps/src/monitoring/http_server.rs]: done, I pointed to the api name and version at the top of the file.

File: stratum-apps/src/monitoring/routes.rs (hunk 6 above)
Comment (gimballock) [reply stratum-apps/src/monitoring/routes.rs]: done

File: stratum-apps/src/monitoring/routes.rs (hunk 5 above)
Comment (gimballock) [reply stratum-apps/src/monitoring/routes.rs]: done

File: integration-tests/lib/prometheus_metrics_assertions.rs [hunk 7]
```diff
@@ -1,30 +1,346 @@
...
+/// Wraps the component's monitoring `SocketAddr` and exposes all
+/// network-touching helpers (raw `/metrics` scrapes, JSON API fetches, and
+/// typed `poll_until_*` helpers) as methods so call sites read as
+/// `api.fetch_typed(...)` / `api.poll_global_sv2_clients_gte(...)` instead of
+/// passing the address into a dozen free functions.
+#[derive(Debug, Clone, Copy)]
+pub struct MonitoringApi {
+    addr: SocketAddr,
+}
+
```
Comment (bit-aloo) [inline integration-tests/lib/prometheus_metrics_assertions.rs:25]: Can you update the MonitoringClient to contain retries, timeout and stuff we discussed during dev sync.

File: integration-tests/lib/prometheus_metrics_assertions.rs [hunk 8]
```diff
@@ -1,30 +1,346 @@
...
+            // pollers are provided for the common cases below.
+            let (status, json) = self.fetch_with_status::<serde_json::Value>(path).await;
+            if (200..300).contains(&status) {
+                if let Some(val) = json.pointer(json_pointer) {
+                    let num = val.as_f64().unwrap_or(0.0);
+                    if num >= min {
+                        return json;
+                    }
+                }
+            }
```
Comment (bit-aloo) [inline integration-tests/lib/prometheus_metrics_assertions.rs]: If we can know rational behind this and maybe improve this?

File: stratum-apps/src/monitoring/http_server.rs (hunk 4 above)
Comment (bit-aloo) [reply stratum-apps/src/monitoring/http_server.rs]: @GitGab19 does these versioning here carries any semantic meaning?

File: integration-tests/lib/prometheus_metrics_assertions.rs (hunk 7 above)
Comment (gimballock) [reply integration-tests/lib/prometheus_metrics_assertions.rs:25]: Done — MonitoringApi now owns retries and request_timeout, and a private http_get_with_status helper centralises URL construction + spawn_blocking + the retry-on-5xx loop. All fetch/poll methods collapsed to one-line wrappers which were then just in-lined directly so these functions could just be removed. The shared panic message now self-describes (path, target type, last status, last body), so we lost no diagnostic info. The default 5s per-request timeout is new — previously minreq calls had no timeout at all, so a wedged server could hang a whole test.

File: integration-tests/lib/prometheus_metrics_assertions.rs (hunk 8 above)
Comment (gimballock) [reply integration-tests/lib/prometheus_metrics_assertions.rs]: This was actually dead-code (now removed) but the json pointer was a way to parse json api responses w/o knowing their types but we have types for all of these responses now anyway so this code was not useful on 2 accounts.

File: integration-tests/lib/prometheus_metrics_assertions.rs (hunk 7 above)
Comment (bit-aloo) [reply integration-tests/lib/prometheus_metrics_assertions.rs:25]: I would prefer a builder struct for this.

Comment (gimballock) [reply integration-tests/lib/prometheus_metrics_assertions.rs:25]: Done — MonitoringApiBuilder now owns retries and request_timeout; MonitoringApi is immutable post-build. MonitoringApi::new(addr) stays as a zero-config shortcut (it just calls builder(addr).build()).

Comment (bit-aloo) [reply integration-tests/lib/prometheus_metrics_assertions.rs:25]: I would want all the construction happen via builder and not any such hacks.

Comment (gimballock) [reply integration-tests/lib/prometheus_metrics_assertions.rs:25]: done

Comment (GitGab19) [reply integration-tests/lib/prometheus_metrics_assertions.rs:25]: What's the benefit of having a builder struct in this case @Shourya742 ?

Comment (bit-aloo) [reply integration-tests/lib/prometheus_metrics_assertions.rs:25]: Because we have two fields, `retries` and `duration`, which take default values, the earlier implementation only allowed overriding them through the `with_*` methods. For example, we had to do `MonitoringClient::new(addr)` and then modify the object using `with_retries` / `with_duration`. That felt a bit counterintuitive, since the object was already constructed and we were mutating it afterward.

With this, we instead use:

```rust
MonitoringClientBuilder::new(addr)
    .retries(retries)
    .duration(duration)
    .build()
```

to construct a configured `MonitoringClient`. We also don’t need to specify `retries` or `duration` unless we want to override the defaults. Once initialization is complete, there is no further mutation of the object.

File: stratum-apps/src/monitoring/routes.rs [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+// Templated full paths (with `{client_id}` placeholder) — used in API
+// listings exposed by the root endpoint and anywhere a human-readable
+// path-with-placeholder is needed.
+pub const CLIENT_BY_ID_PATTERN: &str = "/api/v1/clients/{client_id}";
+pub const CLIENT_CHANNELS_PATTERN: &str = "/api/v1/clients/{client_id}/channels";
+pub const SV1_CLIENT_BY_ID_PATTERN: &str = "/api/v1/sv1/clients/{client_id}";
+
+/// Path for `/api/v1/clients/{id}` with the id substituted in.
+pub fn client_by_id(id: usize) -> String {
+    format!("/api/v1/clients/{id}")
```
Comment (GitGab19) [inline stratum-apps/src/monitoring/routes.rs]:
```suggestion
    format!("{CLIENTS}/{id}")
```

File: stratum-apps/src/monitoring/routes.rs [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+/// Path for `/api/v1/clients/{id}/channels` with the id substituted in.
+pub fn client_channels(id: usize) -> String {
+    format!("/api/v1/clients/{id}/channels")
```
Comment (GitGab19) [inline stratum-apps/src/monitoring/routes.rs]:
```suggestion
    format!("{CLIENTS}/{id}/channels")
```

File: stratum-apps/src/monitoring/routes.rs [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+/// Path for `/api/v1/sv1/clients/{id}` with the id substituted in.
+pub fn sv1_client_by_id(id: usize) -> String {
+    format!("/api/v1/sv1/clients/{id}")
```
Comment (GitGab19) [inline stratum-apps/src/monitoring/routes.rs]:
```suggestion
    format!("{SV1_CLIENTS}/{id}")
```

File: stratum-apps/src/monitoring/http_server.rs (hunk 4 above)
Comment (GitGab19) [reply stratum-apps/src/monitoring/http_server.rs]: For now it's not, except for signalling that APIs could change since they are not stable yet.

File: integration-tests/lib/prometheus_metrics_assertions.rs [hunk 12]
```diff
@@ -158,6 +375,7 @@ pub(crate) fn parse_metric_value<'a, M: Into<Metric<'a>>>(
 }
 
 /// Assert that a metric is present and its value satisfies the given predicate.
+#[track_caller]
```
Comment (GitGab19) [inline integration-tests/lib/prometheus_metrics_assertions.rs]: Is this really useful?

File: integration-tests/lib/prometheus_metrics_assertions.rs [hunk 13]
```diff
@@ -207,6 +420,7 @@ pub fn assert_metric_eq<'a, M: Into<Metric<'a>>>(metrics_text: &str, metric: M,
 /// For a bare-name selector (`Metric::new("name")` or `"name".into()`), this means
 /// the metric name does not appear at all. For a labeled selector, it means no line
 /// with matching labels exists — other series for the same metric name are allowed.
+#[track_caller]
```
Comment (GitGab19) [inline integration-tests/lib/prometheus_metrics_assertions.rs]: Is this really useful?

File: integration-tests/lib/prometheus_metrics_assertions.rs [hunk 14]
```diff
@@ -223,6 +437,7 @@ pub fn assert_metric_not_present<'a, M: Into<Metric<'a>>>(metrics_text: &str, me
 }
 
 /// Assert that at least one exposition line matches the selector.
+#[track_caller]
```
Comment (GitGab19) [inline integration-tests/lib/prometheus_metrics_assertions.rs]: Is this really useful?

**Review by GitGab19 — COMMENTED** [review]
I still haven't tested this yet, but it looks good to me.

Some minor nits.

File: integration-tests/lib/prometheus_metrics_assertions.rs (hunk 14 above)
Comment (gimballock) [reply integration-tests/lib/prometheus_metrics_assertions.rs]: It's not a critical part of the PR, and it's not applied consistently i'll take it out.
We have discussed this [before](https://github.com/stratum-mining/sv2-apps/pull/373#discussion_r3226981719
) on this PR, not sure how to quantify its usefulness.

File: integration-tests/lib/prometheus_metrics_assertions.rs (hunk 13 above)
Comment (gimballock) [reply integration-tests/lib/prometheus_metrics_assertions.rs]: deleted

File: integration-tests/lib/prometheus_metrics_assertions.rs (hunk 12 above)
Comment (gimballock) [reply integration-tests/lib/prometheus_metrics_assertions.rs]: deleted

File: stratum-apps/src/monitoring/routes.rs (hunk 9 above)
Comment (gimballock) [reply stratum-apps/src/monitoring/routes.rs]: done

File: stratum-apps/src/monitoring/routes.rs (hunk 10 above)
Comment (gimballock) [reply stratum-apps/src/monitoring/routes.rs]: done

File: stratum-apps/src/monitoring/routes.rs (hunk 11 above)
Comment (gimballock) [reply stratum-apps/src/monitoring/routes.rs]: done

**Review by bit-aloo — APPROVED** [review]
I am ok to get this in,

**Review by GitGab19 — APPROVED** [review]
